tests/kafkatest/tests/test_upgrade.py

```diff
@@ -0,0 +1,122 @@
+        self.kafka = KafkaService(test_context, num_nodes=3, zk=self.zk, version=KafkaVersion.V_0_8_2_1, topics={self.topic: {
+                                                                    "partitions": 3,
+                                                                    "replication-factor": 3,
+                                                                    "min.insync.replicas": 2}
+                                                                })
+        self.producer_throughput = 1000
+        self.num_producers = 1
+        self.num_consumers = 1
+
+    def setUp(self):
+        self.zk.start()
+        self.kafka.start()
+
+    def test_upgrade(self):
+        """Test upgrade of Kafka broker cluster from 0.8.2 to 0.9.0
+
+        - Start 3 node broker cluster on version 0.8.2
+        - Start producer and consumer in the background
+        - Perform two-phase rolling upgrade
+            - First phase: upgrade brokers to 0.9.0 with inter.broker.protocol.version set to 0.8.2
+            - Second phase: remove inter.broker.protocol.version config with rolling bounce
+        - Finally, validate that every message acked by the producer was consumed by the consumer
+        """
+
+        self.producer = VerifiableProducer(
+            self.test_context, self.num_producers, self.kafka, self.topic,
+            throughput=self.producer_throughput, version=KafkaVersion.V_0_8_2_1)
+        self.consumer = ConsoleConsumer(
+            self.test_context, self.num_consumers, self.kafka, self.topic,
+            consumer_timeout_ms=10000, message_validator=is_int, version=KafkaVersion.V_0_8_2_1)
+
+        # Start background producer and consumer
+        self.producer.start()
+        wait_until(lambda: self.producer.num_acked > 5, timeout_sec=5,
+             err_msg="Producer failed to start in a reasonable amount of time.")
+        self.consumer.start()
+
+        self.logger.info("First pass bounce - rolling upgrade")
+        for node in self.kafka.nodes:
+            self.kafka.stop_node(node)
+            node.version = KafkaVersion.TRUNK
+            node.config[kafka_prop.INTER_BROKER_PROTOCOL_VERSION] = "0.8.2.X"
+            self.kafka.start_node(node)
+
+        self.logger.info("Second pass bounce - remove inter.broker.protocol.version config")
+        for node in self.kafka.nodes:
+            self.kafka.stop_node(node)
+            del node.config[kafka_prop.INTER_BROKER_PROTOCOL_VERSION]
+            self.kafka.start_node(node)
+
+        self.producer.stop()
+        for node in self.consumer.nodes:
+            assert self.consumer.alive(node)
+        self.consumer.wait()
+
+        self.acked = self.producer.acked
+        self.not_acked = self.producer.not_acked
+
+        # Check produced vs consumed
+        self.consumed = self.consumer.messages_consumed[1]
+        self.logger.info("num consumed:  %d" % len(self.consumed))
+        success, msg = self.validate()
+
+        if not success:
```

Comment:
This looks a bit weird -- we have this logic split between `validate` and here. Why two separate `if not success` blocks that mark logs for collection?